pandas/tseries/offsets.py

```diff
@@ -1319,7 +1338,12 @@ def _end_apply_index(self, dtindex):
         else:
             roll = self.n
 
-        base = (base_period + roll).to_timestamp(how='end')
+        with warnings.catch_warnings(record=True):
+            # integer-array addition on PeriodIndex is deprecated,
+            #  but still used internally for performance
```

Comment:
why can't you turn roll into an offset here, rather than catchng the warning? (and all other cases in this module), e.g. since these are not tests, would rather actually fix the issue

Reply:
Using the deprecated behavior amounts to adding two int64 ndarrays.  The non-deprecated version here `base_period + roll * offset` turns `roll*offset` into an object-array, so the whole thing is slow.

I've got a nice solution for this in mind.  This and a few other parts of this PR can be addressed more gracefully after #23031.